ims_usrloc_pcscf:   implementation of db_mode DB_ONLY (value 3)

petermarianF opened the pull request:
In order to support a redundant PCSCF configuration - i.e. a logical PCSCF consists of 2 physical nodes (node1 and node2) some enhancements were introduced for handling of contacts. Redundancy means that SIP messages for a client are normally handled by node1 but in case node1 is not reachable SIP messages are redirected to node2. Of course the DB_ONLY mode must work also for single PCSCF node configuration. Important aspects of this implementation are database integrity - i.e. avoid invalid table entries (for example data which are expired long time ago or have invalid states) - and keeping PCSCF cache in sync with database tables. A wrapper was built for method get_pcontact which tries to find the pcontact in cache and if search is not successful tries to download and insert from db location table - also some effort is added here to find the pcontact if it exists in cache. The contact expiry handler was modified to sync contact expiry in cache with db location entry and in case of real contact expiry sends PUBLISH to SCSCF to let NOTIFY finally delete the contact. An audit for older expired pcontacts was introduced which cares for getting rid of these contacts.
So db queries for this db mode are only supported for Mysql.
Some code was introduced to help registering callbacks for contacts which are inserted into cache when being downloaded from database - for example ims_qos callback as at the time of insertion the message that triggered the original callback registering is long gone.

<!-- Kamailio Pull Request Template -->

<!--
IMPORTANT:
  - for detailed contributing guidelines, read:
    https://github.com/kamailio/kamailio/blob/master/.github/CONTRIBUTING.md
  - pull requests must be done to master branch, unless they are backports
    of fixes from master branch to a stable branch
  - backports to stable branches must be done with 'git cherry-pick -x ...'
  - code is contributed under BSD for core and main components (tm, sl, auth, tls)
  - code is contributed GPLv2 or a compatible license for the other components
  - GPL code is contributed with OpenSSL licensing exception
-->

#### Pre-Submission Checklist
<!-- Go over all points below, and after creating the PR, tick all the checkboxes that apply -->
<!-- All points should be verified, otherwise, read the CONTRIBUTING guidelines from above-->
<!-- If you're unsure about any of these, don't hesitate to ask on sr-dev mailing list -->
- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [ ] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:
<!-- Go over all points below, and after creating the PR, tick the checkboxes that apply -->
- [ ] PR should be backported to stable branches
- [x] Tested changes locally in local branch based on 5.5.4
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description
<!-- see top of this section -->

carstenbock commented:
Thanks for the contribution.

At first review, I did see a few topics:
- please update the XML for the documentation (in the doc/ folder) not the README directly - it would get overwritten, since it is autogenerated
- please make sure to add documentation (in the XML) for the new parameters introduced, in the README only the DB-Mode was added, not the other two new parameters
- instead of commenting code, please remove code. GIT will keep the history, no need to comment

I am no angel myself regarding documentation, but please make sure to keep up with certain standards.

christoph-v-kontron commented:
@carstenbock , @petermarianF 
As far as I know, Peter will issue a few additional PRs soon:
1) db_mode DB_ONLY for S-CSCF
2) Bugfixes for bugs that occur, when db_mode DB_ONLY is active

So I guess, it would be a good idea to issue a separate PR for documentation, as last PR in this series.

Sounds like a plan?

Kr,
Christoph

henningw commented on src/modules/ims_usrloc_pcscf/udomain.c:
```diff
@@ -1077,91 +1113,408 @@ int preload_udomain(db1_con_t* _c, udomain_t* _d)
+        if (_aor->len>0 && _aor->s){
+                LM_INFO("Querying database for P-CSCF contact [%.*s]\n", _aor->len, _aor->s);
+
+    p_location = "SELECT  domain,aor,host,port,protocol,received,received_port,received_proto,rx_session_id,reg_state,expires,socket,service_routes,public_ids,path from location WHERE aor=\"%.*s\""; // AND reg_state=%d";
```
This is not the way you should interact with the DB API. There are actually methods to build a select query. You can find some examples e.g. in the usrloc/dlist.c or in usrloc_db.c files.
This way the module will work with all database drivers that implement the proper functions.

henningw commented on src/modules/ims_usrloc_pcscf/udomain.c:
```diff
@@ -1077,91 +1113,408 @@ int preload_udomain(db1_con_t* _c, udomain_t* _d)
+                querylen = _aor->len;
+        } else {
+                LM_INFO("Querying database for P-CSCF received_host [%.*s] and received_port [%d]\n", contact_info->received_host.len, contact_info->received_host.s, contact_info->received_port);
+                p_location = "SELECT domain,aor,host,port,protocol,received,received_port,received_proto,rx_session_id,reg_state,expires,socket,service_routes,public_ids,path from location WHERE received=\"%.*s\" AND received_port=\"%.*s\"";
```
See above

henningw commented on src/modules/ims_usrloc_pcscf/udomain.c:
```diff
@@ -1077,91 +1113,408 @@ int preload_udomain(db1_con_t* _c, udomain_t* _d)
+pcontact_t* c;
+time_t expired_50secs;
+
+char *p_location = "SELECT  domain,aor,host,port,protocol,received,received_port,received_proto,rx_session_id,reg_state,expires,socket,service_routes,public_ids,path from location WHERE expires < \"%.*s\"";
```
This is not the way you should interact with the DB API. There are actually methods to build a select query. You can find some examples e.g. in the usrloc/dlist.c or usrloc_db files.
This way the module will work with all database drivers that implement the proper functions.

henningw reviewed (changes requested):
Thank you for the PR. I've did longer review today and added several comments.

Many comments are rather small, like removing some commented out code or review and adapt the log levels (which probably are set higher from development). The documentation was already mentioned from Carsten. I've added also some suggestions for naming of parameter.

There is probably one condition regarding memory allocation which needs to be fixed.

The major issue is that you use the DB API incorrect. You should not use raw queries to execute statements on the database. Instead the appropriate select and delete functions should be used. I've added some notes also with pointers to examples.

Just let me know if there are any questions.

henningw commented on src/modules/ims_usrloc_pcscf/ul_callback.c:
```diff
@@ -98,6 +98,39 @@ void destroy_ul_callbacks_list(struct ul_callback* cb) {
+            cbp_registrar->callback = f;
+        }
+        else{
+            if (!(cbp_qos=(struct ul_callback*)shm_malloc(sizeof( struct ul_callback)))) {
```
This variable is causing a conflict with ims_qos module:
/usr/bin/ld: ims_qos_mod.o:./src/modules/ims_qos/./../ims_usrloc_pcscf/../ims_usrloc_pcscf/usrloc.h:246: multiple definition of `cbp_qos'; cdpeventprocessor.o:./src/modules/ims_qos/./../ims_usrloc_pcscf/../ims_usrloc_pcscf/usrloc.h:246: first defined here

Should be renamed

petermarianF commented:
Hi @henningw,
I added one new commit which should resolve all your comments.
In a few minutes I will write an email to ask for further proceeding.

petermarianF commented:
I resolved conflict in src/modules/ims_usrloc_pcscf/doc/ims_usrloc_pcscf_admin.xml
by editing this file and clicking "Mark as resolved"

petermarianF commented:
My resolve action probably did not work as it seems file does not have been saved.
But I copied the changes to my local branch and the README build was fine including Hennings last change. So probably I could add another commit with admin xml

petermarianF commented:
I finally did a rebase to resolve the conflict in the admin xml.
To summarize my changes after first review:
1. removed unnecessary  comments
2. changed INFO/CRIT logs into DBG/ERR logs
3. removed my changes to db_insert_pcontact and db_update_pcontact
4. removed method which deleted pua entry in pua db
5. changed definition of cpb_qos and cpb_registrar to declaration in usrloc.h and moved definition to usrloc.c
6. changed the new db queries to the correct generic approach
7. documented 2 new modparams
8. Free cpb_qos and cpb_registrar  in destroy method

henningw commented on src/modules/ims_usrloc_pcscf/ul_callback.c:
```diff
@@ -98,6 +98,39 @@ void destroy_ul_callbacks_list(struct ul_callback* cb) {
+        else{
+            if (!(cbp_qos=(struct ul_callback*)shm_malloc(sizeof( struct ul_callback)))) {
+                LM_ERR("no more share mem\n");
+                return E_OUT_OF_MEM;
```
I made a mistake, its if/else case so its not executed both

henningw commented on src/modules/ims_usrloc_pcscf/udomain.h:
```diff
@@ -82,5 +83,7 @@ int update_security(udomain_t* _d, security_type _t, security_t* _s, struct pcon
-
+int audit_usrloc_expired_pcontacts(udomain_t* _d);
+int db_load_pcontact(udomain_t* _d, str *_aor, int insert_cache, struct pcontact** _c, pcontact_info_t* contact_info);
+int db_delete_presentityuri_from_pua(str *presentity_uri);
```
This probably can be removed as well, as there is no implementation.

henningw commented on src/modules/ims_usrloc_pcscf/usrloc_db.c:
```diff
@@ -301,7 +307,7 @@ int db_insert_pcontact(struct pcontact* _c)
 	SET_PROPER_NULL_FLAG(_c->rx_session_id, values, LP_RX_SESSION_ID_IDX);
 
-	VAL_DOUBLE(GET_FIELD_IDX(values, LP_REG_STATE_IDX)) = _c->reg_state;
+	VAL_INT(GET_FIELD_IDX(values, LP_REG_STATE_IDX)) = _c->reg_state;
```
Why you are changing this from double to int?

henningw replied:
You are right, it seems to be an error in the module. It probably works because the database converts it automatically.

henningw reviewed (commented):
@petermarianF Thank you for integrating all the comments and sorry for the delay. I've added two more comments, which are probably some minor things.

henningw commented:
Thank you. As there seems to be no more comment from other developers, squashed and merged it. Further changes can be done if necessary directly in the repository.
